File: 2021/d15.hs
import System.IO
import qualified Data.Set as Set
import qualified Data.Array as Array
import Data.List(foldl')
import Data.Char(digitToInt)

import Data.Time

import Util

type Pos = (Int, Int)
data Node = Node
    { riskSum :: Int
    , prev :: Pos
    , level :: Int
    }

main = do
    contents <- readFile "d15.in"
    let input = lines $ trim contents
        dimx = length (head input)
        dimy = length input
        grid =
            Array.array ((0, 0), (dimx - 1, dimy - 1))
                [((x, y),
                  Node { riskSum = maxBound :: Int, prev = (-1, -1), level = digitToInt e }) |
                  (y, l) <- zip [0..] input, (x, e) <- zip [0..] l]
        findPath g pos end =
            go g' ((pos, 0):[]) (Set.fromList [pos]) where
                -- init starting node with a cost of 0
                g' = g Array.// [((0, 0), (g Array.! (0, 0)) { riskSum = 0 })]
                inRange = Array.inRange (Array.bounds g')
                go :: Array.Array Pos Node -> [(Pos, Int)] -> Set.Set Pos -> Int
                go g [] visited = error "Could not find path"
                -- if we keep nodes in the queue they might get outdated when a shorter path is found
                go g (((x, y), pathCost):ps) visited =
                    if (x, y) == end
                    -- end node reached return total risk level
                    then pathCost --riskSum (g Array.! end)
                    else go newG newQ newVisited
                    where
                        -- neighbours we could visit next
                        neighbours = [p | p <- [(x, y - 1), (x + 1, y), (x, y + 1), (x - 1, y)], inRange p]
                        (newG, newQ) = queueOrUpdate g visited ps neighbours
                        newVisited = Set.insert (x, y) visited
                        -- either update visited nodes that now would have a lower riskSum
                        -- or queue them to visit next setting their current riskSum from the current node
                        queueOrUpdate :: Array.Array Pos Node -> Set.Set Pos -> [(Pos, Int)]
                            -> [Pos] -> (Array.Array Pos Node, [(Pos, Int)])
                        queueOrUpdate g v q [] = (g, q)
                        -- use as-pattern syntax (name@(pattern) pattern matches on the pattern as well
                        -- as binding the whole matched structure to name) to match on the record fields
                        -- AND bind the whole record to a name "nn"
                        -- queueOrUpdate g v q ((pos, nn@(Node { riskSum = rs, level = nlvl })):ns) =
                        queueOrUpdate g v q (pos:ps) =
                            if Set.member pos visited
                            -- visited notes should already have the optimal cost so just skip them
                            then queueOrUpdate g v q ps
                            -- not visited -> update cost and queue
                            else
                                if oldCost > newCost
                                -- new path is cheaper -> update and queue
                                then queueOrUpdate gridUpdated v (insertSorted (pos, newCost) q) ps
                                -- NOTE: part1 worked on 2nd run (after fixing compile errors only 1 run where
                                -- I forgot to update the node in the grid if it wasn't visited below)
                                -- UPDATE1: it only worked for the example since we carried around outdated node
                                -- instances (due to immutability they won't be updated)
                                -- UPDATE2: only was updating costs on visited nodes which by the design of the
                                -- algo should already have the optimal costs, now update non-visited nodes
                                -- and queue them if cost is smaller (since initial cost is maxInt this works
                                -- for non-queued as well); this works since we only want to get to our target
                                -- (it would not work if we wanted the shortest paths to all nodes in the grid)
                                -- NOTE: technically should also check if queued already and update it in the q
                                -- queue item where position is based on riskSum
                                else queueOrUpdate g v q ps
                            where nn = g Array.! pos
                                  oldCost = riskSum nn
                                  nlvl = level nn
                                  newCost = pathCost + nlvl
                                  nodeUpdated = nn { prev = (x, y), riskSum = newCost }
                                  gridUpdated = (g Array.// [(pos, nodeUpdated)])
                        insertSorted i [] = [i]
                        insertSorted i (y:ys) =
                            let irs = snd i
                                crs = snd y in
                            if irs < crs
                            then i:y:ys
                            else y : (insertSorted i ys)

    -- repeat grid 4 times to the rigth and below, where the risklvl increases every time
    -- we go to the right or down
    start <- getCurrentTime
    let fiveXGrid =
            Array.array ((0, 0), (5 * dimx - 1, 5 * dimy - 1))
                [((x + dimx * xx, y + dimy * yy), --((r - 1 + xx + yy) `mod` 9) + 1) |
                  Node { riskSum = maxBound :: Int, prev = (-1, -1)
                       -- NOTE: level increases every step to the right or down with wrap around >9 to 1
                       -- -1 to get it into the 0-8 range then mod 9 so we get wrapping
                       -- add 1 to get it back to 1-9
                       , level = ((r - 1 + xx + yy) `mod` 9) + 1 }) |
                  (y, l) <- zip [0..] input, (x, e) <- zip [0..] l, let r = digitToInt e,
                  xx <- [0..4], yy <- [0..4]]
        printArray arr =
            unlines [unwords [show (arr Array.! (x, y)) | x <- [10..19]] | y <- [0..9]]



    -- putStrLn $ show $ findPath grid (0, 0) (dimx - 1, dimy - 1)
    -- putStrLn $ printArray fiveXGrid
    -- COMPILED version takes 728.1910445s WOW
    putStrLn $ show $ findPath fiveXGrid (0, 0) (5 * dimx - 1, 5 * dimy - 1)
    stop <- getCurrentTime
    print $ diffUTCTime stop start
    -- python: 1.5s (with binsearch; without 6.7s)
    -- julia 0.4s with same binsearch as python (own impl); julia 5s (with stdlib bin search 0.5s)
    -- mikefarquhar's rust solution: 30ms
    -- toakushi's go solution: 184ms Dijkstra (A* 214ms)
    -- tpatetl's nodejs solution: 224s
    -- heyitsmattwade's nodejs solution: 1.5s (using a heap/prioqueue library)
    -- IT WAS ACTUALLY IMPOSSIBLE TO FIND A JS IMPLEMENTATION THAT DID NOT use
    -- A LIBRARY!!_)!!_)_!_!__!_!_!! HOW? what's the point of AoC then
    -- 3/5 used a lib for dijkstra which IS THE WHOLE CODING TASK!?!??!
    -- can't use binsearch with Haskell since the q is a list and iterating over the list
    -- is what's slow
    -- Haskell 728.1910445s
    -- F#: ~13s (with dotnet fsi d15.fsx command which apparently gets compiled)
    -- (^ also no binsearch, and using the same algo as Haskell, the ONLY difference is that
    --  F# uses the mutable array whereas Haskell uses the immutable Data.Array implementation)

File: 2021/d15.cpp
#include <cassert>
#include <iostream>
#include <chrono>
#include <fstream>
#include <array>
#include <unordered_set>
#include <vector>
#include <set>
#include <string>
#include <limits>
#include <tuple>

// NOTE: converted from Python using ChatGPT
//       then fixed all errors/outdated syntax
// all versions exactly follow the algorithm used by d15.py to be comparable
// -> this C++ version could be optimized, e.g. by
//    - using a continguous 1D array for the grid
//    - same for visited with vec<bool>
//    - std::priority_queue for the queue instead of manual binsearch
// g++ -std=c++17 -O3 -march=native -o d15_cpp d15.cpp

struct Node {
    int level;
    int risk_sum;
    std::pair<int,int> prev;

    Node(int lvl, int risk) : level(lvl), risk_sum(risk), prev({-1,-1}) {}
};

using Pos = std::pair<int,int>;

// Custom hasher struct
struct PosHash {
    size_t operator()(const Pos& p) const noexcept {
        return std::hash<int>{}(p.first) ^ (std::hash<int>{}(p.second) << 1);
    }
};

// Binary search insertion index (descending cost queue)
size_t binsearch(const std::vector<std::pair<Pos,int>>& arr, int x) {
    if (arr.empty()) {
        return 0;
    }
    size_t first = 0;
    size_t last = arr.size() - 1;

    while (first < last) {
        size_t mid = first + (last - first) / 2;
        int val = arr[mid].second;
        if (val == x) return mid;
        else if (val > x) {
            if (first == mid) return (arr[last].second < x) ? last : last + 1;
            first = mid;
        } else {
            last = mid;
        }
    }
    return 0;
}

// Find shortest path using the same queue algorithm
int find_path(std::vector<std::vector<Node>>& grid, Pos start, Pos end_pos) {
    size_t dimy = grid.size();
    size_t dimx = grid[0].size();

    std::vector<std::pair<Pos,int>> q = { {start, 0} };
    // NOTE: unordered_set + custom hasher, since we need a hash-based set
    //       to be comparable with Python/Mojo
    // std::unordered_set<Pos, PosHash> visited{};
    std::set<Pos> visited{};
    visited.insert(start);

    while (!q.empty()) {
        auto [pos, path_cost] = q.back();
        q.pop_back();
        int x = pos.first;
        int y = pos.second;

        if (pos == end_pos) return path_cost;

        visited.insert(pos);

        const std::array<Pos, 4> neighbours{{
            {x, y-1}, {x+1, y}, {x, y+1}, {x-1, y} }};
        for (auto npos : neighbours) {
            int nx = npos.first;
            int ny = npos.second;
            if (nx < 0 || nx >= (int)dimx || ny < 0 || ny >= (int)dimy) continue;
            if (visited.count(npos)) continue;

            Node& neighbour_node = grid[ny][nx];
            int new_cost = path_cost + neighbour_node.level;
            if (new_cost < neighbour_node.risk_sum) {
                neighbour_node.risk_sum = new_cost;
                neighbour_node.prev = pos;

                size_t idx = binsearch(q, new_cost);
                q.insert(q.begin() + idx, {npos, new_cost});
            }
        }
    }
    return -1; // unreachable
}

int main() {
    std::ifstream infile("d15.in");
    std::string line;
    std::vector<std::string> contents;
    while (std::getline(infile, line)) {
        if (!line.empty()) contents.push_back(line);
    }

    // Build initial grid
    std::vector<std::vector<Node>> grid;
    for (auto& l : contents) {
        std::vector<Node> row;
        for (char c : l) {
            row.emplace_back(c - '0', std::numeric_limits<int>::max());
        }
        grid.emplace_back(std::move(row));
    }

    int part1 = find_path(grid, {0,0}, { (int)grid[0].size()-1, (int)grid.size()-1 });
    std::cout << "Part1: " << part1 << std::endl;

    // Build 5x expanded grid
    std::vector<std::vector<Node>> five_x_grid;
    for (int yy = 0; yy < 5; ++yy) {
        for (auto& l : contents) {
            std::vector<Node> row;
            for (int xx = 0; xx < 5; ++xx) {
                for (char c : l) {
                    int new_level = (c - '0' - 1 + xx + yy) % 9 + 1;
                    row.emplace_back(new_level, std::numeric_limits<int>::max());
                }
            }
            five_x_grid.emplace_back(std::move(row));
        }
    }

    auto start_time = std::chrono::high_resolution_clock::now();
    int part2 = find_path(five_x_grid, {0,0}, { (int)five_x_grid[0].size()-1, (int)five_x_grid.size()-1 });
    auto end_time = std::chrono::high_resolution_clock::now();

    std::cout << "Part2: " << part2 << std::endl;

    double took_ms = std::chrono::duration<double, std::milli>(end_time - start_time).count();
    std::cout << "Took: " << took_ms << " ms" << std::endl;

    return 0;
}
